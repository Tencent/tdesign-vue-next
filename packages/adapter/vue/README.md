# 适配过程记录

## 1. vue
1. instance

```ts
props: innerInstance?.$props
```
## 2. hooks
- useIcon