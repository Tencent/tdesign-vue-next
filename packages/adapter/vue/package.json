{
  "name": "@td/adapter-vue",
  "version": "1.0.0",
  "description": "",
  "author": "",
  "license": "ISC",
  "keywords": [],
  "main": "index.ts",
  "module": "index.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "devDependencies": {
    "@td/ts-config": "workspace:^"
  }
}
