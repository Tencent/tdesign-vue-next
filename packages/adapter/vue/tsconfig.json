{
  "extends": "@td/ts-config/vue3.json",
  "compilerOptions": {
    // 应该设置为 true 的，但因 adapter 导出 type，所以设置为 false 了 
    "verbatimModuleSyntax": false,
  }
}
