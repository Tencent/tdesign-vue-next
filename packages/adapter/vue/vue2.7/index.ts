import Vue, { h as H, getCurrentInstance as getCurrentInstanceInner } from 'vue';
import type { Ref, VNode, VueConstructor } from 'vue';
import { clone, lowerFirst, merge, omit } from 'lodash-es';
import Teleport from './components/Teleport';

function getCurrentInstance() {
  const innerInstance = getCurrentInstanceInner()?.proxy;
  const instance = innerInstance
    ? {
        ...innerInstance,
        props: innerInstance.$props,
        vnode: innerInstance.$vnode,
        emit: innerInstance.$emit.bind(innerInstance),
        slots: innerInstance.$scopedSlots,
      }
    : undefined;

  return instance;
}

function getVNode(node: VNode) {
  return {
    ...node,
    props: node.componentOptions?.propsData,
    children: { default: () => node.componentOptions?.children },
  };
}

function cloneVNode(node: VNode, extraProps: Record<string, string | number | Array<string | number>>) {
  return merge(clone(node), extraProps);
}

export interface JSXRenderContext {
  defaultNode?: VNode | string;
  params?: Record<string, any>;
  // 是否不打印 LOG
  silent?: boolean;
}

export type OptionsType = VNode | JSXRenderContext | string;

function isVNode(obj: OptionsType) {
  const vNode = H('span', '');
  const VNode = vNode.constructor;
  return obj instanceof VNode;
}

const Transition = Vue.component('Transition');
const TransitionGroup = Vue.component('TransitionGroup');

function createApp(rootComp: VueConstructor, rootProps: Record<string, any> = {}) {
  const comp = Vue.extend(rootComp);
  const ins = new comp({ propsData: rootProps });
  const mount = (el: string | HTMLElement) => {
    ins.$mount(el);
    return ins;
  };
  return { ...ins, mount };
}

function pluginInstall(app: VueConstructor, plugin: Plugin, ...paths: string[]) {
  let lastObject = app.prototype;
  for (let i = 0; i < paths.length - 1; i++) {
    const path = paths[i];
    if (!path) {
      throw new Error(`Path ${i} is empty`);
    }
    if (!(path in lastObject)) {
      lastObject[path] = {};
    }
    lastObject = lastObject[path];
  }
  const lastPath = paths[paths.length - 1];
  if (!lastPath) {
    throw new Error('Last path is empty');
  }
  lastObject[lastPath] = plugin;
}

// vue23:?? 临时写的方法，待讨论
function createElement(type: string, props: Record<string, any> = {}, children?: any) {
  const innerData = omit(props, ['key', 'ref', 'class', 'style']);
  const event: Record<string, any> = {};
  const data: Record<string, any> = {};

  for (const key in innerData) {
    if (key.startsWith('on')) {
      event[lowerFirst(key.slice(2))] = innerData[key];
    } else {
      data[key] = innerData[key];
    }
  }

  return H(type, {
    key: props.key,
    ref: props.ref,
    class: props.class,
    style: props.style,
    attrs: data,
    props: data,
    on: event,
  }, children);
}

const isVue3 = false;
const isVue2 = true;

export * from 'vue';
export {
  getVNode,
  getCurrentInstance,
  H,
  cloneVNode,
  isVNode,
  Teleport,
  Transition,
  TransitionGroup,
  createApp,
  createElement,
  pluginInstall,
  isVue3,
  isVue2,
};

export default Vue;
