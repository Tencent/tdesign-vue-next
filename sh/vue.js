const fs = require('fs');
const path = require('path');

// 定义源文本的正则表达式
const src_regex = /from 'vue';/g;

// 定义要搜索的目录
const dirs = ['packages/components/common/src', 'packages/components/vue3/src'];

// 定义处理文件的函数
function processFiles(directory) {
    fs.readdirSync(directory, { withFileTypes: true }).forEach(file => {
        const filePath = path.join(directory, file.name);
        if (file.isDirectory()) {
            processFiles(filePath);
        } else if (file.isFile() && (filePath.endsWith('.js') || filePath.endsWith('.ts') || filePath.endsWith('.tsx'))) {
            let content = fs.readFileSync(filePath, 'utf8');
            content = content.replace(src_regex, "from '@td/adapter-vue';");
            fs.writeFileSync(filePath, content, 'utf8');
        }
    });
}

// 遍历目录并处理文件
dirs.forEach(dir => processFiles(dir));