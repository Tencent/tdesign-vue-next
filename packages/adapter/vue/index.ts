import * as Vue from '@adapter/vue';

export * from '@adapter/vue';

export default Vue;
