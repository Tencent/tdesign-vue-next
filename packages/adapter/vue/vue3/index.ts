import * as Vue from 'vue';
import { h as H, getCurrentInstance } from 'vue';
import type { App, Plugin, VNode } from 'vue';

function getVNode(node: VNode) {
  return node;
}

function pluginInstall(app: App, plugin: Plugin, ...paths: string[]) {
  let lastObject = app.config.globalProperties;
  for (let i = 0; i < paths.length - 1; i++) {
    const path = paths[i];
    if (!path) {
      throw new Error(`Path ${i} is empty`);
    }
    if (!(path in lastObject)) {
      lastObject[path] = {};
    }
    lastObject = lastObject[path];
  }
  const lastPath = paths[paths.length - 1];
  if (!lastPath) {
    throw new Error('Last path is empty');
  }
  lastObject[lastPath] = plugin;
}

const createElement = H;

const isVue3 = true;
const isVue2 = false;

export * from 'vue';
export {
  getCurrentInstance,
  getVNode,
  createElement,
  H,
  pluginInstall,
  isVue3,
  isVue2,
};

export default Vue;
