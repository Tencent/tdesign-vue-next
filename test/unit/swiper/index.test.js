import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import Swiper, { SwiperItem } from '@/src/swiper/index.ts';

describe('Swiper', () => {
  describe(':props', () => {
    it('', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper>
              <SwiperItem>1</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.exists()).toBe(true);
    });
    it(':animation fade', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper animation={'fade'}>
              <SwiperItem>1</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':autoplay', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper autoplay>
              <SwiperItem>1</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':defaultCurrent', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper defaultCurrent={2}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':direction', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper direction={'vertical'}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':duration', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper duration={3000}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':height', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper height={600}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':interval', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper interval={3000}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':loop', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper loop={true}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':navigation', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper
              navigation={{
                placement: 'outside',
                showSlideBtn: 'hover',
                size: 'large',
                type: 'dots-bar',
              }}
            >
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':theme', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper theme={'dark'}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    it(':type', () => {
      const wrapper = mount({
        render() {
          return (
            <Swiper type={'card'}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  // test events
  describe('@event', () => {
    it('@change', async () => {
      const fn = jest.fn();
      const wrapper = mount({
        render() {
          return (
            <Swiper onChange={fn}>
              <SwiperItem>1</SwiperItem>
              <SwiperItem>2</SwiperItem>
            </Swiper>
          );
        },
      });
      await nextTick();
      const swiper = wrapper.findComponent(Swiper);
      swiper.find('.t-swiper__arrow-right').trigger('click');
      await nextTick();
      expect(fn).toHaveBeenCalled();
    });
  });
});
