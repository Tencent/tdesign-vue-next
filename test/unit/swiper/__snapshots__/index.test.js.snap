// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swiper :props :animation fade 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content t-swiper-fade"
    >
      <div
        class="t-swiper__container"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :autoplay 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :defaultCurrent 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-200%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :direction 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside t-swiper--vertical"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(0, -0%, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :duration 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :height 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :interval 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :loop 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :navigation 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--outside t-swiper--large"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :theme 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content"
    >
      <div
        class="t-swiper__container"
        style="transform: translate3d(-0%, 0px, 0px);"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;

exports[`Swiper :props :type 1`] = `
<div
  class="t-swiper"
>
  <div
    class="t-swiper__wrap t-swiper--inside"
  >
    <div
      class="t-swiper__content t-swiper-card"
    >
      <div
        class="t-swiper__container"
      >
        
        
      </div>
    </div>
    <ul
      class="t-swiper__navigation t-swiper__navigation-bars"
    >
      
      
    </ul>
    <!---->
  </div>
</div>
`;
